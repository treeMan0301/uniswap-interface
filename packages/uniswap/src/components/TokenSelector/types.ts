import { GqlResult } from 'uniswap/src/data/types'
import { CurrencyInfo } from 'uniswap/src/features/dataApi/types'
import { TokenSelectorFlow } from 'uniswap/src/features/transactions/transfer/types'
import { UniverseChainId } from 'uniswap/src/types/chains'
import { FiatNumberType } from 'utilities/src/format/types'

export type TokenOption = {
  currencyInfo: CurrencyInfo
  quantity: number | null // float representation of balance, returned by data-api
  balanceUSD: Maybe<number>
}

export type TokenOptionsHookType = (
  address: string,
  chainFilter: UniverseChainId | null,
  searchFilter?: string,
) => GqlResult<TokenOption[] | undefined>

export type TokenOptionsWithChainFilterHookType = (
  address: string,
  chainFilter: UniverseChainId,
  searchFilter?: string,
) => GqlResult<TokenOption[] | undefined>

export type TokenOptionsWithBalanceOnlySearchHookType = (
  chainFilter: UniverseChainId | null,
  searchFilter: string | null,
  isBalancesOnlySearch: boolean,
) => GqlResult<TokenSection[]>

export type OnSelectCurrency = (
  currency: CurrencyInfo,
  section: SuggestedTokenSection | TokenSection,
  index: number,
) => void

export type TokenSection = {
  title: string
  data: TokenOption[]
  rightElement?: JSX.Element
}

export type SuggestedTokenSection = {
  title: string
  data: TokenOption[][]
  rightElement?: JSX.Element
}

export type TokenSelectorListSections = Array<SuggestedTokenSection | TokenSection>

export type TokenWarningDismissedHook = (currencyId: Maybe<string>) => {
  tokenWarningDismissed: boolean
  dismissWarningCallback: () => void
}

export type TokenSectionsForSwap = {
  activeAccountAddress: string
  chainFilter: UniverseChainId | null
  usePopularTokensOptionsHook: (address: string, chainFilter: UniverseChainId) => GqlResult<TokenOption[] | undefined>
  usePortfolioTokenOptionsHook: (
    address: string,
    chainFilter: UniverseChainId | null,
    searchFilter?: string | undefined,
  ) => GqlResult<TokenOption[] | undefined>
}

export type TokenSectionsForSwapInput = TokenSectionsForSwap

export type TokenSectionsForSwapOutput = Omit<TokenSectionsForSwap, 'usePortfolioTokenOptionsHook'> & {
  useFavoriteTokensOptionsHook: (
    address: string,
    chainFilter: UniverseChainId | null,
  ) => GqlResult<TokenOption[] | undefined>
  useCommonTokensOptionsHook: (
    address: string,
    chainFilter: UniverseChainId | null,
  ) => GqlResult<TokenOption[] | undefined>
}

export type TokenSectionsForSend = Omit<TokenSectionsForSwap, 'usePopularTokensOptionsHook'>

export type ConvertFiatAmountFormattedCallback = (
  fromAmount: Maybe<string | number>,
  numberType: FiatNumberType,
  placeholder?: string | undefined,
) => string

export type FilterCallbacksHookType = (
  chainId: UniverseChainId | null,
  flow: TokenSelectorFlow,
) => {
  chainFilter: UniverseChainId | null
  searchFilter: string | null
  onChangeChainFilter: (newChainFilter: UniverseChainId | null) => void
  onClearSearchFilter: () => void
  onChangeText: (newSearchFilter: string) => void
}
