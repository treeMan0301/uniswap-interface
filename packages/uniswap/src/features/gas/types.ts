import { SerializedError } from '@reduxjs/toolkit'
import { FetchError } from 'uniswap/src/data/apiClients/FetchError'

export enum GasSpeed {
  Normal = 'normal',
  Fast = 'fast',
  Urgent = 'urgent',
}

export enum FeeType {
  Legacy = 'legacy',
  Eip1559 = 'eip1559',
}

interface GasFeeResponseBase {
  type: FeeType
  gasLimit: string
  gasFee: {
    normal: string
    fast: string
    urgent: string
  }
}

interface GasFeeResponseEip1559 extends GasFeeResponseBase {
  type: FeeType.Eip1559
  maxFeePerGas: {
    normal: string
    fast: string
    urgent: string
  }
  maxPriorityFeePerGas: {
    normal: string
    fast: string
    urgent: string
  }
}

interface GasFeeResponseLegacy extends GasFeeResponseBase {
  type: FeeType.Legacy
  gasPrice: {
    normal: string
    fast: string
    urgent: string
  }
}

export type TransactionLegacyFeeParams = {
  gasPrice: string
  gasLimit: string
}

export type TransactionEip1559FeeParams = {
  maxFeePerGas: string
  maxPriorityFeePerGas: string
  gasLimit: string
}

// GasFeeResponse is the type that comes directly from the Gas Service API
export type GasFeeResponse = GasFeeResponseEip1559 | GasFeeResponseLegacy

export type GasFeeResult = {
  value?: string
  isLoading: boolean
  error: SerializedError | FetchError | Error | null
  params?: TransactionLegacyFeeParams | TransactionEip1559FeeParams
}

export type FormattedUniswapXGasFeeInfo = {
  approvalFeeFormatted?: string
  wrapFeeFormatted?: string
  swapFeeFormatted: string
  preSavingsGasFeeFormatted: string
  inputTokenSymbol?: string
}
