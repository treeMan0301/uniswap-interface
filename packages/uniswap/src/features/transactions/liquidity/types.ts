// eslint-disable-next-line no-restricted-imports
import { ProtocolVersion } from '@uniswap/client-pools/dist/pools/v1/types_pb'
import { Currency, CurrencyAmount, Token } from '@uniswap/sdk-core'
import { IncreaseLPPositionRequest } from 'uniswap/src/data/tradingApi/__generated__'
import { ValidatedPermit, ValidatedTransactionRequest } from 'uniswap/src/features/transactions/swap/utils/trade'

export interface LiquidityAction {
  currency0Amount: CurrencyAmount<Currency>
  currency1Amount: CurrencyAmount<Currency>
  liquidityToken: Token
}

export type LiquidityTxAndGasInfo = IncreasePositionTxAndGasInfo
export type ValidatedLiquidityTxContext = ValidatedIncreasePositionTxAndGasInfo

export function isValidLiquidityTxContext(
  liquidityTxContext: LiquidityTxAndGasInfo | unknown,
): liquidityTxContext is ValidatedLiquidityTxContext {
  // Validation fn prevents/future-proofs typeguard against illicit casts
  return validateLiquidityTxContext(liquidityTxContext) !== undefined
}

interface BaseRequiredLiquidityTxContextFields {
  approvalError: false
}

interface BaseLiquidityTxAndGasInfo {
  protocolVersion: ProtocolVersion
  action: LiquidityAction
  approvalError: boolean
  approveToken0Request: ValidatedTransactionRequest | undefined
  approveToken1Request: ValidatedTransactionRequest | undefined
  approvePositionTokenRequest: ValidatedTransactionRequest | undefined
  permit: ValidatedPermit | undefined
  revocationTxRequest: ValidatedTransactionRequest | undefined
}

export interface IncreasePositionTxAndGasInfo extends BaseLiquidityTxAndGasInfo {
  unsigned: boolean
  increasePositionRequestArgs: IncreaseLPPositionRequest | undefined
  txRequest: ValidatedTransactionRequest | undefined
}

export type ValidatedIncreasePositionTxAndGasInfo = Required<IncreasePositionTxAndGasInfo> &
  BaseRequiredLiquidityTxContextFields &
  (
    | {
        unsigned: true
        permit: ValidatedPermit
        txRequest: undefined
      }
    | {
        unsigned: false
        permit: undefined
        txRequest: ValidatedTransactionRequest
      }
  )

function validateLiquidityTxContext(
  liquidityTxContext: LiquidityTxAndGasInfo | unknown,
): ValidatedLiquidityTxContext | undefined {
  if (!isLiquidityTx(liquidityTxContext)) {
    return undefined
  }

  if (!liquidityTxContext.approvalError && liquidityTxContext.action) {
    const { approvalError } = liquidityTxContext
    const { action, txRequest, unsigned, permit } = liquidityTxContext

    if (unsigned) {
      if (!permit) {
        return undefined
      }
      return { ...liquidityTxContext, action, approvalError, unsigned, txRequest: undefined, permit }
    } else if (txRequest) {
      return { ...liquidityTxContext, action, approvalError, unsigned, txRequest, permit: undefined }
    }
  }

  return undefined
}

function isLiquidityTx(liquidityTxContext: unknown): liquidityTxContext is LiquidityTxAndGasInfo {
  return (
    typeof liquidityTxContext === 'object' &&
    liquidityTxContext !== null &&
    'action' in liquidityTxContext &&
    'protocolVersion' in liquidityTxContext
  )
}
