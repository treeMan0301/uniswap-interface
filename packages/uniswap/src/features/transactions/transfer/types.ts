export enum TokenSelectorFlow {
  Swap,
  Transfer,
}
