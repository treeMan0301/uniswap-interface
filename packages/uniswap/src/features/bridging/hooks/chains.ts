import { useMemo } from 'react'
import { useTradingApiSwappableTokensQuery } from 'uniswap/src/data/apiClients/tradingApi/useTradingApiSwappableTokensQuery'
import { ChainId } from 'uniswap/src/data/tradingApi/__generated__'
import {
  NATIVE_ADDRESS_FOR_TRADING_API,
  toTradingApiSupportedChainId,
} from 'uniswap/src/features/transactions/swap/utils/tradingApi'
import { UniverseChainId } from 'uniswap/src/types/chains'

const FALLBACK_NUM_CHAINS = 8

export function useNumBridgingChains(): number {
  const { data: bridgingTokens } = useTradingApiSwappableTokensQuery({
    params: {
      tokenIn: NATIVE_ADDRESS_FOR_TRADING_API,
      tokenInChainId: ChainId._1,
    },
  })

  const chainSet = useMemo(() => new Set(bridgingTokens?.tokens.map((t) => t.chainId)), [bridgingTokens])

  return chainSet.size + 1 ?? FALLBACK_NUM_CHAINS
}

export function useIsBridgingChain(chainId: UniverseChainId): boolean {
  const { data: bridgingTokens } = useTradingApiSwappableTokensQuery({
    params: {
      tokenIn: NATIVE_ADDRESS_FOR_TRADING_API,
      tokenInChainId: ChainId._1,
    },
  })

  const chainSet = useMemo(() => new Set(bridgingTokens?.tokens.map((t) => t.chainId)), [bridgingTokens])

  const chainIdForTradingApi = toTradingApiSupportedChainId(chainId)
  return chainIdForTradingApi !== undefined && chainSet.has(chainIdForTradingApi)
}
