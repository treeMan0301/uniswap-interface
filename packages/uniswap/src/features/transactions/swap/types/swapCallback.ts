import { CurrencyAmount, Currency } from "@uniswap/sdk-core"
import { SignerMnemonicAccountMeta } from "uniswap/src/features/accounts/types"
import { ValidatedSwapTxContext } from "uniswap/src/features/transactions/swap/contexts/SwapTxContext"

export interface SwapCallbackParams {
  account: SignerMnemonicAccountMeta
  swapTxContext: ValidatedSwapTxContext
  currencyInAmountUSD: Maybe<CurrencyAmount<Currency>>
  currencyOutAmountUSD: Maybe<CurrencyAmount<Currency>>
  isAutoSlippage: boolean
  onSubmit: () => void
  onFailure: () => void
  txId?: string
  isFiatInputMode?: boolean
}

export type SwapCallback = (params: SwapCallbackParams) => void
