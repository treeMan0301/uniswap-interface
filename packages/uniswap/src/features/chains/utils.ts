import { BigNumber, BigNumberish } from '@ethersproject/bignumber'
import { ChainId } from '@uniswap/sdk-core'
import {
  GQL_MAINNET_CHAINS_MUTABLE,
  GQL_TESTNET_CHAINS_MUTABLE,
  UNIVERSE_CHAIN_INFO,
} from 'uniswap/src/constants/chains'
import { PollingInterval } from 'uniswap/src/constants/misc'
import { Chain } from 'uniswap/src/data/graphql/uniswap-data-api/__generated__/types-and-hooks'
import { FeatureFlags } from 'uniswap/src/features/gating/flags'
import { useFeatureFlag } from 'uniswap/src/features/gating/hooks'
import {
  COMBINED_CHAIN_IDS,
  InterfaceGqlChain,
  NetworkLayer,
  SUPPORTED_CHAIN_IDS,
  SUPPORTED_TESTNET_CHAIN_IDS,
  UniverseChainId,
} from 'uniswap/src/types/chains'

export function toGraphQLChain(chainId: ChainId | number): Chain | undefined {
  switch (chainId) {
    case ChainId.MAINNET:
      return Chain.Ethereum
    case ChainId.SEPOLIA:
      return Chain.EthereumSepolia
    case ChainId.ARBITRUM_ONE:
      return Chain.Arbitrum
    case ChainId.OPTIMISM:
      return Chain.Optimism
    case ChainId.POLYGON:
      return Chain.Polygon
    case ChainId.BASE:
      return Chain.Base
    case ChainId.BNB:
      return Chain.Bnb
    case ChainId.AVALANCHE:
      return Chain.Avalanche
    case ChainId.CELO:
      return Chain.Celo
    case ChainId.BLAST:
      return Chain.Blast
    case ChainId.WORLDCHAIN:
      return Chain.Worldchain
    case ChainId.ZORA:
      return Chain.Zora
    case ChainId.ZKSYNC:
      return Chain.Zksync
    case ChainId.ASTROCHAIN_SEPOLIA:
      return Chain.AstrochainSepolia
  }
  return undefined
}

// Some code from the web app uses chainId types as numbers
// This validates them as coerces into SupportedChainId
export function toSupportedChainId(chainId?: BigNumberish): UniverseChainId | null {
  const ids = COMBINED_CHAIN_IDS

  if (!chainId || !ids.map((c) => c.toString()).includes(chainId.toString())) {
    return null
  }
  return parseInt(chainId.toString(), 10) as UniverseChainId
}

export function chainIdToHexadecimalString(chainId: UniverseChainId): string {
  return BigNumber.from(chainId).toHexString()
}

export function hexadecimalStringToInt(hex: string): number {
  return parseInt(hex, 16)
}

export function isL2ChainId(chainId?: UniverseChainId): boolean {
  return chainId !== undefined && UNIVERSE_CHAIN_INFO[chainId].networkLayer === NetworkLayer.L2
}

export function isMainnetChainId(chainId?: UniverseChainId): boolean {
  return chainId === UniverseChainId.Mainnet || chainId === UniverseChainId.Sepolia
}

export function fromGraphQLChain(chain: Chain | undefined): UniverseChainId | null {
  switch (chain) {
    case Chain.Ethereum:
      return UniverseChainId.Mainnet
    case Chain.Arbitrum:
      return UniverseChainId.ArbitrumOne
    case Chain.EthereumSepolia:
      return UniverseChainId.Sepolia
    case Chain.Optimism:
      return UniverseChainId.Optimism
    case Chain.Polygon:
      return UniverseChainId.Polygon
    case Chain.Base:
      return UniverseChainId.Base
    case Chain.Bnb:
      return UniverseChainId.Bnb
    case Chain.Blast:
      return UniverseChainId.Blast
    case Chain.Avalanche:
      return UniverseChainId.Avalanche
    case Chain.Celo:
      return UniverseChainId.Celo
    case Chain.Worldchain:
      return UniverseChainId.WorldChain
    case Chain.Zora:
      return UniverseChainId.Zora
    case Chain.Zksync:
      return UniverseChainId.Zksync
    case Chain.AstrochainSepolia:
      return UniverseChainId.AstrochainSepolia
  }

  return null
}

export function getPollingIntervalByBlocktime(chainId?: UniverseChainId): PollingInterval {
  return isMainnetChainId(chainId) ? PollingInterval.Fast : PollingInterval.LightningMcQueen
}

export function fromMoonpayNetwork(moonpayNetwork: string | undefined): UniverseChainId | undefined {
  switch (moonpayNetwork) {
    case Chain.Arbitrum.toLowerCase():
      return UniverseChainId.ArbitrumOne
    case Chain.Optimism.toLowerCase():
      return UniverseChainId.Optimism
    case Chain.Polygon.toLowerCase():
      return UniverseChainId.Polygon
    case Chain.Bnb.toLowerCase():
      return UniverseChainId.Bnb
    // Moonpay still refers to BNB chain as BSC so including both BNB and BSC cases
    case 'bsc':
      return UniverseChainId.Bnb
    case Chain.Base.toLowerCase():
      return UniverseChainId.Base
    case Chain.Avalanche.toLowerCase():
      return UniverseChainId.Avalanche
    case Chain.Celo.toLowerCase():
      return UniverseChainId.Celo
    case undefined:
      return UniverseChainId.Mainnet
    default:
      return undefined
  }
}

export function fromUniswapWebAppLink(network: string | null): UniverseChainId | null {
  switch (network) {
    case Chain.Ethereum.toLowerCase():
      return UniverseChainId.Mainnet
    case Chain.EthereumSepolia.toLowerCase():
      return UniverseChainId.Sepolia
    case Chain.Arbitrum.toLowerCase():
      return UniverseChainId.ArbitrumOne
    case Chain.Optimism.toLowerCase():
      return UniverseChainId.Optimism
    case Chain.Polygon.toLowerCase():
      return UniverseChainId.Polygon
    case Chain.Base.toLowerCase():
      return UniverseChainId.Base
    case Chain.Bnb.toLowerCase():
      return UniverseChainId.Bnb
    case Chain.Blast.toLowerCase():
      return UniverseChainId.Blast
    case Chain.Avalanche.toLowerCase():
      return UniverseChainId.Avalanche
    case Chain.Celo.toLowerCase():
      return UniverseChainId.Celo
    case Chain.Worldchain.toLowerCase():
      return UniverseChainId.WorldChain
    case Chain.Zora.toLowerCase():
      return UniverseChainId.Zora
    case Chain.Zksync.toLowerCase():
      return UniverseChainId.Zksync
    case Chain.AstrochainSepolia.toLowerCase():
      return UniverseChainId.AstrochainSepolia
    default:
      throw new Error(`Network "${network}" can not be mapped`)
  }
}

export function toUniswapWebAppLink(chainId: UniverseChainId): string | null {
  switch (chainId) {
    case UniverseChainId.Mainnet:
      return Chain.Ethereum.toLowerCase()
    case UniverseChainId.Sepolia:
      return Chain.EthereumSepolia.toLowerCase()
    case UniverseChainId.ArbitrumOne:
      return Chain.Arbitrum.toLowerCase()
    case UniverseChainId.Optimism:
      return Chain.Optimism.toLowerCase()
    case UniverseChainId.Polygon:
      return Chain.Polygon.toLowerCase()
    case UniverseChainId.Base:
      return Chain.Base.toLowerCase()
    case UniverseChainId.Bnb:
      return Chain.Bnb.toLowerCase()
    case UniverseChainId.Blast:
      return Chain.Blast.toLowerCase()
    case UniverseChainId.Avalanche:
      return Chain.Avalanche.toLowerCase()
    case UniverseChainId.Celo:
      return Chain.Celo.toLowerCase()
    case UniverseChainId.WorldChain:
      return Chain.Worldchain.toLowerCase()
    case UniverseChainId.Zora:
      return Chain.Zora.toLowerCase()
    case UniverseChainId.Zksync:
      return Chain.Zksync.toLowerCase()
    case UniverseChainId.AstrochainSepolia:
      return Chain.AstrochainSepolia.toLowerCase()
    default:
      throw new Error(`ChainID "${chainId}" can not be mapped`)
  }
}

type ActiveChainIdFeatureFlags = UniverseChainId.WorldChain

export function filterChainIdsByFeatureFlag(featureFlaggedChainIds: {
  [UniverseChainId.WorldChain]: boolean
}): UniverseChainId[] {
  return COMBINED_CHAIN_IDS.filter((chainId) => {
    return featureFlaggedChainIds[chainId as ActiveChainIdFeatureFlags] ?? true
  })
}

// Used to feature flag chains. If a chain is not included in the object, it is considered enabled by default.
export function useFeatureFlaggedChainIds(): UniverseChainId[] {
  // You can use the useFeatureFlag hook here to enable/disable chains based on feature flags.
  // Example: [ChainId.BLAST]: useFeatureFlag(FeatureFlags.BLAST)
  // IMPORTANT: Don't forget to also update getEnabledChainIdsSaga

  const worldChainEnabled = useFeatureFlag(FeatureFlags.WorldChain)

  return filterChainIdsByFeatureFlag({
    [UniverseChainId.WorldChain]: worldChainEnabled,
  })
}

export function getEnabledChains({
  isTestnetModeEnabled,
  featureFlaggedChainIds,
  connectedWalletChainIds,
}: {
  isTestnetModeEnabled: boolean
  featureFlaggedChainIds: UniverseChainId[]
  connectedWalletChainIds?: UniverseChainId[]
}): {
  chains: UniverseChainId[]
  gqlChains: InterfaceGqlChain[]
  defaultChainId: UniverseChainId
  isTestnetModeEnabled: boolean
} {
  if (isTestnetModeEnabled) {
    const supportedTestnetChainIds = SUPPORTED_TESTNET_CHAIN_IDS.filter(
      (chainId) =>
        featureFlaggedChainIds.includes(chainId) &&
        (connectedWalletChainIds ? connectedWalletChainIds.includes(chainId) : true),
    )

    return {
      chains: supportedTestnetChainIds,
      gqlChains: GQL_TESTNET_CHAINS_MUTABLE,
      defaultChainId: UniverseChainId.Sepolia as UniverseChainId,
      isTestnetModeEnabled,
    }
  }

  const supportedChainIds = SUPPORTED_CHAIN_IDS.filter(
    (chainId) =>
      featureFlaggedChainIds.includes(chainId) &&
      (connectedWalletChainIds ? connectedWalletChainIds.includes(chainId) : true),
  )

  return {
    chains: supportedChainIds,
    gqlChains: GQL_MAINNET_CHAINS_MUTABLE,
    defaultChainId: UniverseChainId.Mainnet as UniverseChainId,
    isTestnetModeEnabled,
  }
}
