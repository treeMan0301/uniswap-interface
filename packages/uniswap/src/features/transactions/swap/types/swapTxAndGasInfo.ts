import { Routing, NullablePermit, CreateSwapRequest, OrderRequest } from "uniswap/src/data/tradingApi/__generated__/index"
import { GasFeeResult } from "uniswap/src/features/gas/types"
import { ClassicTrade, IndicativeTrade, UniswapXTrade } from "uniswap/src/features/transactions/swap/types/trade"
import { isClassic, isUniswapX } from "uniswap/src/features/transactions/swap/utils/routing"
import { ValidatedTransactionRequest } from "uniswap/src/features/transactions/swap/utils/trade"
import { GasFeeEstimates } from "uniswap/src/features/transactions/types/transactionDetails"

export type SwapTxAndGasInfo = ClassicSwapTxAndGasInfo | UniswapXSwapTxAndGasInfo
export type ValidatedSwapTxContext = ValidatedClassicSwapTxAndGasInfo | ValidatedUniswapXSwapTxAndGasInfo

export function isValidSwapTxContext(swapTxContext: SwapTxAndGasInfo): swapTxContext is ValidatedSwapTxContext {
  // Validation fn prevents/future-proofs typeguard against illicit casts
  return validateSwapTxContext(swapTxContext) !== undefined
}

export type SwapGasFeeEstimation = {
  swapEstimates?: GasFeeEstimates
  approvalEstimates?: GasFeeEstimates
}

export type UniswapXGasBreakdown = {
  classicGasUseEstimateUSD?: string
  approvalCost?: string
  wrapCost?: string
  inputTokenSymbol?: string
}

interface BaseSwapTxAndGasInfo {
  routing: Routing
  trade?: ClassicTrade | UniswapXTrade
  indicativeTrade: IndicativeTrade | undefined
  approvalError: boolean
  permitData: NullablePermit | undefined
  permitDataLoading: boolean | undefined
  approveTxRequest: ValidatedTransactionRequest | undefined
  revocationTxRequest: ValidatedTransactionRequest | undefined
  gasFee: GasFeeResult
}

export interface ClassicSwapTxAndGasInfo extends BaseSwapTxAndGasInfo {
  routing: Routing.CLASSIC
  trade?: ClassicTrade
  txRequest?: ValidatedTransactionRequest
  swapRequestArgs: CreateSwapRequest | undefined
  permitSignature: string | undefined
  gasFeeEstimation: SwapGasFeeEstimation
}

export interface UniswapXSwapTxAndGasInfo extends BaseSwapTxAndGasInfo {
  routing: Routing.DUTCH_V2
  trade: UniswapXTrade
  wrapTxRequest: ValidatedTransactionRequest | undefined
  orderParams?: OrderRequest
  gasFeeBreakdown: UniswapXGasBreakdown
}
interface BaseRequiredSwapTxContextFields {
  approvalError: false
  gasFee: ValidatedGasFeeResult
}

type ValidatedClassicSwapTxAndGasInfo = Required<ClassicSwapTxAndGasInfo> & BaseRequiredSwapTxContextFields
type ValidatedUniswapXSwapTxAndGasInfo = Required<UniswapXSwapTxAndGasInfo> & BaseRequiredSwapTxContextFields

function validateSwapTxContext(swapTxContext: SwapTxAndGasInfo): ValidatedSwapTxContext | undefined {
  const gasFee = validateGasFeeResult(swapTxContext.gasFee)
  if (!gasFee) {
    return undefined
  }

  if (!swapTxContext.approvalError && swapTxContext.trade) {
    const { approvalError } = swapTxContext
    if (isClassic(swapTxContext) && swapTxContext.trade && swapTxContext.txRequest) {
      const { trade, txRequest } = swapTxContext
      return { ...swapTxContext, trade, txRequest, approvalError, gasFee }
    } else if (isUniswapX(swapTxContext) && swapTxContext.orderParams) {
      const { trade, orderParams } = swapTxContext
      return { ...swapTxContext, trade, gasFee, approvalError, orderParams }
    }
  }
  return undefined
}

type ValidatedGasFeeResult = GasFeeResult & { value: string; error: null }
function validateGasFeeResult(gasFee: GasFeeResult): ValidatedGasFeeResult | undefined {
  if (gasFee.value === undefined || gasFee.error) {
    return undefined
  }
  return { ...gasFee, value: gasFee.value, error: null }
}

