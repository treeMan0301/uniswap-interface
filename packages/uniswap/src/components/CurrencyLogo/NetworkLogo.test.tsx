import {
  NetworkLogo,
  TransactionSummaryNetworkLogo,
} from 'uniswap/src/components/CurrencyLogo/NetworkLogo'
import { render } from 'uniswap/src/test/test-utils'
import { ChainId } from 'uniswap/src/types/chains'

jest.mock('uniswap/src/constants/chains', () => {
  const actualChains = jest.requireActual('uniswap/src/constants/chains')

  const mockedChainInfo = {
    ...actualChains.UNIVERSE_CHAIN_INFO,
    ['chainWithoutLogo']: {
      logo: undefined, // no logo - for testing
    },
  }

  return {
    ...actualChains,
    UNIVERSE_CHAIN_INFO: mockedChainInfo,
  }
})

describe('NetworkLogo', () => {
  const ACTUAL_CHAIN_INFO = jest.requireActual('uniswap/src/constants/chains').UNIVERSE_CHAIN_INFO

  it('renders without error', () => {
    const tree = render(<NetworkLogo chainId={ChainId.Base} />)

    expect(tree).toMatchSnapshot()
  })

  it('renders logo when the chain info has a logo', () => {
    Object.keys(ACTUAL_CHAIN_INFO).forEach((chainId) => {
      const tree = render(<NetworkLogo chainId={chainId as unknown as ChainId} />)
      expect(tree.toJSON()).toBeTruthy()
    })
  })

  it('renders null when chain info has no logo', () => {
    const tree = render(<NetworkLogo chainId={'chainWithoutLogo' as unknown as ChainId} />)

    expect(tree.toJSON()).toBeNull()
  })
})

describe(TransactionSummaryNetworkLogo, () => {
  it('renders without error', () => {
    const tree = render(<TransactionSummaryNetworkLogo chainId={ChainId.Base} />)

    expect(tree).toMatchSnapshot()
  })
})
