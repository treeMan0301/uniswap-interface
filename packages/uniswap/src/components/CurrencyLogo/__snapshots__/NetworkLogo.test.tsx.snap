// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NetworkLogo renders without error 1`] = `
<View
  style={
    {
      "borderBottomColor": "#131313",
      "borderBottomLeftRadius": 6,
      "borderBottomRightRadius": 6,
      "borderBottomWidth": 0,
      "borderLeftColor": "#131313",
      "borderLeftWidth": 0,
      "borderRightColor": "#131313",
      "borderRightWidth": 0,
      "borderStyle": "solid",
      "borderTopColor": "#131313",
      "borderTopLeftRadius": 6,
      "borderTopRightRadius": 6,
      "borderTopWidth": 0,
      "flexDirection": "column",
      "overflow": "hidden",
    }
  }
  testID="network-logo"
>
  <div
    className="css-view-175oi2r r-flexBasis-1mlwlqe r-overflow-1udh08x r-zIndex-417010"
    dir={null}
    style={
      {
        "height": "20px",
        "width": "20px",
      }
    }
  >
    <div
      className="css-view-175oi2r r-backgroundColor-1niwhzg r-backgroundPosition-vvn4in r-backgroundRepeat-u6sd8q r-bottom-1p0dtai r-height-1pi2tsx r-left-1d2f490 r-position-u8s1d r-right-zchlnj r-top-ipm5af r-width-13qz1uu r-zIndex-1wyyakw r-backgroundSize-ehq7j7"
      dir={null}
      suppressHydrationWarning={true}
    />
  </div>
</View>
`;

exports[`TransactionSummaryNetworkLogo renders without error 1`] = `
<View
  style={
    {
      "borderBottomColor": "#131313",
      "borderBottomLeftRadius": 6.8999999999999995,
      "borderBottomRightRadius": 6.8999999999999995,
      "borderBottomWidth": 1.5,
      "borderLeftColor": "#131313",
      "borderLeftWidth": 1.5,
      "borderRightColor": "#131313",
      "borderRightWidth": 1.5,
      "borderStyle": "solid",
      "borderTopColor": "#131313",
      "borderTopLeftRadius": 6.8999999999999995,
      "borderTopRightRadius": 6.8999999999999995,
      "borderTopWidth": 1.5,
      "flexDirection": "column",
      "overflow": "hidden",
    }
  }
  testID="network-logo"
>
  <div
    className="css-view-175oi2r r-flexBasis-1mlwlqe r-overflow-1udh08x r-zIndex-417010"
    dir={null}
    style={
      {
        "height": "20px",
        "width": "20px",
      }
    }
  >
    <div
      className="css-view-175oi2r r-backgroundColor-1niwhzg r-backgroundPosition-vvn4in r-backgroundRepeat-u6sd8q r-bottom-1p0dtai r-height-1pi2tsx r-left-1d2f490 r-position-u8s1d r-right-zchlnj r-top-ipm5af r-width-13qz1uu r-zIndex-1wyyakw r-backgroundSize-ehq7j7"
      dir={null}
      suppressHydrationWarning={true}
    />
  </div>
</View>
`;
